### PR TITLE
Reduce jitted function overhead

### PR DESCRIPTION
**Pull request by ricardoV94**
This reduces the overhead on the benchmarked numba function from ~10us to 2.5us on my machine, with trust_input=True

It will hopefully go further down when we remove deprecated function stuff like output_subset and dict returns

----
📚 Documentation preview 📚: https://pytensor--1101.org.readthedocs.build/en/1101/

### CODE REVIEW

File: tests/link/numba/test_basic.py [hunk 1]
```diff
@@ -882,3 +882,20 @@ def test_cache_warning_suppressed():
...
     x_test = np.random.uniform(size=5)
     np.testing.assert_allclose(fn(x_test), scipy.special.psi(x_test) * 2)
+
+
+@pytest.mark.parametrize("mode", ("default", "trust_input", "direct"))
+def test_function_overhead(mode, benchmark):
+    x = pt.vector("x")
+    out = pt.exp(x)
+
+    fn = function([x], out, mode="NUMBA")
```
Comment (Ch0ronomato) [inline tests/link/numba/test_basic.py:892]: Should we also test this on torch? Do you want me to do that?

File: tests/link/numba/test_basic.py [hunk 2]
```diff
@@ -882,3 +882,20 @@ def test_cache_warning_suppressed():
...
+    if mode == "trust_input":
+        fn.trust_input = True
+    elif mode == "direct":
+        fn = fn.vm.jit_fn
+
+    test_x = np.zeros(1000)
+    assert np.sum(fn(test_x)) == 1000
+
+    benchmark(fn, test_x)
```
Comment (Ch0ronomato) [inline tests/link/numba/test_basic.py:901]: Does benchmark fail if some amount of time is elapsed?

File: pytensor/link/pytorch/linker.py [hunk 3]
```diff
@@ -25,7 +14,16 @@ def fgraph_convert(self, fgraph, input_storage, storage_map, **kwargs):
...
 
-        return torch.compile(fn)
+        from pytensor.link.pytorch.dispatch import pytorch_typify
+
+        inner_fn = torch.compile(fn)
+
+        # Torch does not accept numpy inputs and may return GPU objects
+        def fn(*inputs, inner_fn=inner_fn):
+            outs = inner_fn(*(pytorch_typify(inp) for inp in inputs))
+            return tuple(out.cpu() for out in outs)
```
Comment (Ch0ronomato) [inline pytensor/link/pytorch/linker.py:24]: I have a few PR's open that might cobble this fyi.

File: pytensor/link/pytorch/linker.py [hunk 4]
```diff
@@ -25,7 +14,16 @@ def fgraph_convert(self, fgraph, input_storage, storage_map, **kwargs):
     def jit_compile(self, fn):
         import torch
 
-        return torch.compile(fn)
+        from pytensor.link.pytorch.dispatch import pytorch_typify
+
+        inner_fn = torch.compile(fn)
+
+        # Torch does not accept numpy inputs and may return GPU objects
+        def fn(*inputs, inner_fn=inner_fn):
```
Comment (Ch0ronomato) [inline pytensor/link/pytorch/linker.py:22]: Maybe just use the closure scoped function?

**Review by Ch0ronomato — COMMENTED** [review]
Mostly looked at the torch linker and some q's around the tests.

File: tests/link/numba/test_basic.py (hunk 2 above)
Comment (ricardoV94) [reply tests/link/numba/test_basic.py:901]: We have a job that should show a warning but I don't think it's working. It at least allows us to do git bisect or something like that over commits to see if performance was dropped

File: pytensor/link/pytorch/linker.py (hunk 4 above)
Comment (ricardoV94) [reply pytensor/link/pytorch/linker.py:22]: Wdym?

File: pytensor/link/pytorch/linker.py (hunk 3 above)
Comment (ricardoV94) [reply pytensor/link/pytorch/linker.py:24]: Whichever gets merged first, we can solve the conflicts after

File: tests/link/numba/test_basic.py (hunk 1 above)
Comment (ricardoV94) [reply tests/link/numba/test_basic.py:892]: I think there's much more going on with Torch that we may want to think about more carefully which sort of graph to benchmark and also cpu/gpu. Let's open an issue for when the torch backend is a bit more established?

File: pytensor/link/numba/linker.py [hunk 5]
```diff
@@ -1,26 +1,15 @@
...
     """A `Linker` that JIT-compiles NumPy-based operations using Numba."""
 
-    def output_filter(self, var: "Variable", out: Any) -> Any:
-        if not isinstance(var, np.ndarray) and isinstance(
-            var.type, pytensor.tensor.TensorType
-        ):
-            return var.type.filter(out, allow_downcast=True)
-
-        return out
-
```
Comment (ricardoV94) [inline pytensor/link/numba/linker.py:23]: This was actually wrong, It probably was meant to be `if not isinstance(out, np.ndarray)`, so it always triggered `var.type.fliter` which was quite slow.

Anyway, if numba is returning a non-array where we expected an array it means something is wrong in our dispatch, and we should fix it there.

File: pytensor/link/pytorch/linker.py (hunk 4 above)
Comment (Ch0ronomato) [reply pytensor/link/pytorch/linker.py:80]: I think I usually see us just reference `inner_fn` without declaring an optional param. I think I've seen both but whatever works.

**Review by jessegrabowski — APPROVED** [review]
lgtm, I left a dumb comment that you can ignore (the gc stuff was re-implemented in the last commit).

File: pytensor/link/basic.py [hunk 6]
```diff
@@ -701,34 +696,7 @@ def make_all(self, input_storage=None, output_storage=None, storage_map=None):
             compute_map, nodes, input_storage, output_storage, storage_map
         )
 
-        computed, last_user = gc_helper(nodes)
-
-        if self.allow_gc:
```
[the remark this answers is not in the document]
Comment (ricardoV94) [reply pytensor/link/basic.py:706]: It doesn't do anything for jitted functions where you don't control intermediate allocations

**Pull request merged**